chain-signatures/node/src/indexer_eth/indexer_eth_helios.rs

```diff
@@ -0,0 +1,230 @@
+        to: Address,
+        data: Bytes,
+        block_number: u64,
+    ) -> anyhow::Result<Bytes> {
+        let helios_client = self.client.clone();
+
+        // Build a base tx *without* the sentinel max gas
+        let mut tx = TransactionRequest::default()
+            .from(from)
+            .to(to)
+            .input(alloy::rpc::types::TransactionInput::both(data.clone()));
+
+        // 1) Estimate
+        let est = helios_client
+            .estimate_gas(&tx, BlockId::Number(BlockNumberOrTag::Number(block_number)))
+            .await
+            .unwrap_or(3_000_000u64); // fallback
+
+        // 2) Add 20% buffer, but keep < 16,777,216
+        let mut gas = (est as f64 * 1.2) as u64;
+        if gas > 16_777_216 {
+            gas = 16_777_216;
+        }
+
+        // 3) Apply gas limit
+        tx = tx.gas_limit(gas);
+
+        // 4) Execute the call
+        helios_client
+            .call(&tx, BlockId::Number(BlockNumberOrTag::Number(block_number)))
+            .await
+            .map_err(|err| anyhow::anyhow!("Failed to call: {err:?}"))
+    }
+
+    async fn get_latest_block_number(&self) -> anyhow::Result<u64> {
+        let Some(block) = self
+            .fetch_block(BlockId::Number(BlockNumberOrTag::Latest))
+            .await
+        else {
+            return Err(anyhow::anyhow!("Latest block not found"));
+        };
+        Ok(block.header.number)
+    }
+
+    async fn get_transaction_receipt(
+        &self,
+        tx_hash: alloy::primitives::B256,
+    ) -> anyhow::Result<Option<alloy::rpc::types::TransactionReceipt>> {
+        self.client
+            .get_transaction_receipt(tx_hash)
+            .await
+            .map_err(|err| anyhow::anyhow!("Failed to get transaction receipt: {err:?}"))
+    }
+}
+
+pub async fn build_client(eth: EthConfig) -> anyhow::Result<HeliosEthereumClient> {
+    let Ok(network) = Network::from_str(eth.network.as_str()) else {
+        return Err(anyhow::anyhow!("Network input incorrect: {}", eth.network));
+    };
+    let client: EthereumClient = {
+        let builder = match EthereumClientBuilder::new()
+            .network(network)
+            .consensus_rpc(&eth.consensus_rpc_http_url)
+        {
+            Ok(builder) => builder,
+            Err(err) => {
+                return Err(anyhow::anyhow!("Failed to build consensus RPC: {err:?}"));
+            }
+        };
+
+        let builder = match builder.execution_rpc(&eth.execution_rpc_http_url) {
+            Ok(builder) => builder,
+            Err(err) => {
+                return Err(anyhow::anyhow!("Failed to build execution RPC: {err:?}"));
+            }
+        };
+
+        match builder
+            .data_dir(PathBuf::from(&eth.helios_data_path))
+            .with_file_db()
+            .build()
+        {
+            Ok(client) => client,
+            Err(err) => {
+                return Err(anyhow::anyhow!("Failed to build Helios client: {err:?}"));
+            }
+        }
+    };
```

Comment:
just use anyhow context with these:
```suggestion
    let client = EthereumClientBuilder::new()
        .network(network)
        .consensus_rpc(&eth.consensus_rpc_http_url)
        .context("failed to build consensus rpc")?
        .execution_rpc(&eth.execution_rpc_http_url)
        .context("failed to build execution rpc")?
        .data_dir(PathBuf::from(&eth.helios_data_path))
        .with_file_db()
        .build()
        .context("failed to build helios client")?;
```